data/transactions/logic/eval.go

```diff
@@ -2571,6 +2571,76 @@ func opSetByte(cx *evalContext) {
 	cx.stack = cx.stack[:prev]
 }
 
+func extract(x []byte, start, length int) (out []byte, err error) {
+	out = x
+	end := start + length
+	if start > len(x) || end > len(x) {
+		err = errors.New("extract range beyond length of string")
+		return
+	}
+	out = x[start:end]
+	err = nil
+	return
+}
+
+func opExtract(cx *evalContext) {
+	last := len(cx.stack) - 1
+	start := cx.program[cx.pc+1]
+	length := cx.program[cx.pc+2]
+	// Shortcut: if length is 0, take bytes from start index to the end
+	l := int(length)
+	if length == 0 {
+		l = len(cx.stack[last].Bytes) - int(start)
+	}
+	cx.stack[last].Bytes, cx.err = extract(cx.stack[last].Bytes, int(start), l)
+}
+
+func opExtract3(cx *evalContext) {
+	last := len(cx.stack) - 1 // length
```

Comment:
Could just be me, but using `last` as an argument index and `start` as a slicing index confused me. What do you think about using names such as `lengthIdx`, `startIdx`, and `bytesIdx`? This would also make it easier to read line 2608 without having to refer back to what `pprev` is.

Reply:
I think your suggestion makes more sense -- will change the indices to have `*Idx`.